sympy/polys/monomials.py

```diff
@@ -50,44 +59,102 @@ def itermonomials(variables, max_degree, min_degree = 0):
 
         >>> sorted(itermonomials([x, y], 2, 1), key=monomial_key('grlex', [y, x]))
         [x, y, x**2, x*y, y**2]
 
 
+    Case II:: `max_degrees` and `min_degrees` are both lists.
+    =========================================================
+    If max_degrees = [d_1, ..., d_n] and min_degrees = [e_1, ..., e_n],
+    the number of monomials generated is:
+
+        (d_1 - e_1 + 1) * ... * (d_n - e_n + 1)
+
+    Example
+    =======
+
+    Let us generate all monomials `monom` in variables `x`, and `y`
+    such that [1, 2][i] <= degree_list(monom)[i] <= [2, 4][i], i = 0, 1 ::
+
+        >>> from sympy import symbols
+        >>> from sympy.polys.monomials import itermonomials
+        >>> from sympy.polys.orderings import monomial_key
+        >>> from itertools import product
+        >>> from sympy.core import Mul
+        >>> from sympy.abc import x, y
+
+        >>> sorted(itermonomials([x, y], [2, 4], [1, 2]), reverse=True, key=monomial_key('lex', [x, y]))
+        [x**2*y**4, x**2*y**3, x**2*y**2, x*y**4, x*y**3, x*y**2]
     """
-    if max_degree < 0 or min_degree > max_degree:
-        return set()
-    if not variables or max_degree == 0:
-        return {S(1)}
-    # Force to list in case of passed tuple or other incompatible collection
-    variables = list(variables) + [S(1)]
-    if all(variable.is_commutative for variable in variables):
-        monomials_list_comm = []
-        for item in combinations_with_replacement(variables, max_degree):
-            powers = dict()
-            for variable in variables:
-                powers[variable] = 0
-            for variable in item:
-                if variable != 1:
-                    powers[variable] += 1
-            if sum(powers.values()) >= min_degree:
-                monomials_list_comm.append(Mul(*item))
-        return set(monomials_list_comm)
+    n = len(variables)
+    if is_sequence(max_degrees):
+        if len(max_degrees) != n:
+            raise ValueError('Argument sizes do not match')
+        if min_degrees is None:
+            min_degrees = [0]*n
+        elif not is_sequence(min_degrees):
+            raise ValueError('min_degrees is not a list')
+        else:
+            if len(min_degrees) != n:
+                raise ValueError('Argument sizes do not match')
+            if any(i < 0 for i in min_degrees):
+                raise ValueError("min_degrees can't contain negative numbers")
+        total_degree = False
+    else:
+        max_degree = max_degrees
+        if max_degree < 0:
+            raise ValueError("max_degrees can't be negative")
+        if min_degrees is None:
+            min_degree = 0
+        else:
+            if min_degrees < 0:
+                raise ValueError("min_degrees can't be negative")
+            min_degree = min_degrees
+        total_degree = True
+    if total_degree:
+        if min_degree > max_degree:
+            yield set()
+        if not variables or max_degree == 0:
+            yield {S(1)}
```

Comment:
Here a single item is returned (not a set) and then return added to indicate StopIteration.

    yield S(1)
    return

Reply:
Again, thanks for the suggestion. It should be a bit faster like this.

Reply:
This should yield the number `S(1)`, not the set `{S(1)}`. The tests below should also have `set` instead of `next`.